### PR TITLE
enforce early cancellation fee on cancel

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 1]
```diff
@@ -185,24 +201,9 @@ contract KeeperRegistryLogic is KeeperRegistryBase {
...
-      if (cancellationFee > amountLeft) {
-        cancellationFee = amountLeft;
-      }
-    }
-    uint96 amountToWithdraw = amountLeft - cancellationFee;
-
-    s_upkeep[id].balance = 0;
-    s_ownerLinkBalance = s_ownerLinkBalance + cancellationFee;
-
+    uint96 amountToWithdraw = s_upkeep[id].balance;
```
Comment (FelixFan1992) [inline contracts/src/v0.8/dev/KeeperRegistryLogic.sol:204]: now when withdrawing, transfer all balance to user

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 2]
```diff
@@ -1450,125 +1453,6 @@ describe('KeeperRegistryDev', () => {
         registration = await registry.getUpkeep(id)
         assert.equal(0, registration.balance.toNumber())
       })
-
```
Comment (FelixFan1992) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:1453]: move these cancellation fee related tests to cancelUpkeep testsuite

File: contracts/test/v0.8/dev/KeeperRegistryDev.test.ts [hunk 3]
```diff
@@ -1696,6 +1580,28 @@ describe('KeeperRegistryDev', () => {
     describe('when called by the admin', async () => {
       const delay = 50
 
+      it('reverts if called again by the admin', async () => {
```
Comment (FelixFan1992) [inline contracts/test/v0.8/dev/KeeperRegistryDev.test.ts:1583]: rearrange these 2 revert test cases so all revert cases are on top.

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 4]
```diff
@@ -152,12 +152,13 @@ contract KeeperRegistryLogic is KeeperRegistryBase {
    * @dev Called through KeeperRegistry main contract
    */
   function cancelUpkeep(uint256 id) external {
-    uint64 maxValid = s_upkeep[id].maxValidBlocknumber;
+    Upkeep memory upkeep = s_upkeep[id];
+    uint64 maxValid = upkeep.maxValidBlocknumber;
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:156]: Done

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 5]
```diff
@@ -166,6 +167,21 @@ contract KeeperRegistryLogic is KeeperRegistryBase {
     s_upkeep[id].maxValidBlocknumber = uint64(height);
     s_upkeepIDs.remove(id);
 
+    // charge the cancellation fee if the minUpkeepSpend is not met
+    uint96 minUpkeepSpend = s_storage.minUpkeepSpend;
+    uint96 amountLeft = upkeep.balance;
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:172]: Done

File: contracts/src/v0.8/dev/KeeperRegistryLogic.sol [hunk 6]
```diff
@@ -166,6 +167,21 @@ contract KeeperRegistryLogic is KeeperRegistryBase {
...
+    uint96 amountSpent = upkeep.amountSpent;
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply contracts/src/v0.8/dev/KeeperRegistryLogic.sol:173]: Done

**Review by infiloop2 — APPROVED** [review]
lgtm!

**Pull request merged**